File: board-back/src/main/java/com/example/boardback/service/BoardService.java
package com.example.boardback.service;

import org.springframework.http.ResponseEntity;

import com.example.boardback.dto.request.board.PatchBoardRequestDto;
import com.example.boardback.dto.request.board.PostBoardRequestDto;
import com.example.boardback.dto.request.board.PostCommentRequestDto;
import com.example.boardback.dto.response.board.DeleteBoardResponseDto;
import com.example.boardback.dto.response.board.GetBoardResponseDto;
import com.example.boardback.dto.response.board.GetCommentListResponseDto;
import com.example.boardback.dto.response.board.GetFavoriteListResponseDto;
import com.example.boardback.dto.response.board.GetLatestBoardListResponseDto;
import com.example.boardback.dto.response.board.GetTop3BoardListResponseDto;
import com.example.boardback.dto.response.board.IncreaseViewCountResponseDto;
import com.example.boardback.dto.response.board.PatchBoardResponseDto;
import com.example.boardback.dto.response.board.PostBoardResponseDto;
import com.example.boardback.dto.response.board.PostCommentResponseDto;
import com.example.boardback.dto.response.board.PutFavoriteResponseDto;

public interface BoardService {

  ResponseEntity<? super GetBoardResponseDto> getBoard(Integer boardNumber);
  ResponseEntity<? super GetFavoriteListResponseDto> getFavoriteList(Integer boardNumber);
  ResponseEntity<? super GetCommentListResponseDto> getCommentList(Integer boardNumber);
  ResponseEntity<? super GetLatestBoardListResponseDto> getLatestBoardList();
  ResponseEntity<? super GetTop3BoardListResponseDto> getTop3BoardList();

  
  ResponseEntity<? super PostBoardResponseDto> postBoard(PostBoardRequestDto dto, String email); 
  ResponseEntity<? super PostCommentResponseDto> postComment(PostCommentRequestDto dto, Integer boardNumber, String email); 

  ResponseEntity<? super PutFavoriteResponseDto> putFavorite(Integer boardNumber, String email);
  ResponseEntity<? super PatchBoardResponseDto> patchBoard(PatchBoardRequestDto dto, Integer boardNumber, String email);

  ResponseEntity<? super IncreaseViewCountResponseDto> inceaseViewCount(Integer boardNumber);

  ResponseEntity<? super DeleteBoardResponseDto> deleteBoard(Integer boardNumber, String eamil);

}

File: board-back-nest/src/modules/board/board.service.ts
import { Injectable } from '@nestjs/common';
import { BoardRepository, CommentRepository, ImageRepository, UserRepository } from 'modules/data-access/repository';
import { PatchBoardRequestDto, PostBoardRequestDto, PostCommentRequestDto } from './dto/request';
import { GetBoardResponseDto, PatchBoardResponseDto, PostBoardResponseDto, PostCommentResponseDto } from './dto/response';

@Injectable()
export class BoardService {
  
  constructor(
    private readonly userRepository: UserRepository,
    private readonly boardRepository: BoardRepository,
    private readonly imageRepository: ImageRepository,
    private readonly commentRepository: CommentRepository
  ) {}

  async postBoard(dto:PostBoardRequestDto, email:string): Promise<PostBoardResponseDto> {

    const isExistUser = await this.userRepository.existsByEmail(email);
    if(!isExistUser) PostBoardResponseDto.noExistUser();

    const boardEntity = this.boardRepository.create(dto, email);
    await this.boardRepository.save(boardEntity);

    const { boardImageList } = dto;
    const { boardNumber } = boardEntity;
    const imageEntities = this.imageRepository.createAll(boardImageList, boardNumber);
    await this.imageRepository.saveAll(imageEntities);

    return PostBoardResponseDto.success();

  }

  async postComment(dto:PostCommentRequestDto, boardNumber: number, email:string):Promise<PostCommentResponseDto> {
    
    const isExistUser = await this.userRepository.existsByEmail(email);
    if(!isExistUser) PostCommentResponseDto.noExistBoard();

    const boardEntity = await this.boardRepository.findByBoardNumber(boardNumber);
    if(!boardEntity) PostCommentResponseDto.noExistBoard();

    const commentEntity = this.commentRepository.create(dto, boardNumber, email);
    await this.commentRepository.save(commentEntity);
 
    boardEntity.commentCount++;  // 댓글 카운트 1 증가 // private 지정안되서 여기서 사용가능
    await this.boardRepository.save(boardEntity);

    return PostCommentResponseDto.success();
  }

  async getBoard(boardNumber: number): Promise<GetBoardResponseDto> {
    const resultSet = await this.boardRepository.getBoard(boardNumber);
    if(!resultSet) GetBoardResponseDto.noExistBoard();

    const imageEntities = await this.imageRepository.findByBoardNumber(boardNumber);

    return GetBoardResponseDto.success(resultSet, imageEntities);
  }

  async patchBoard(dto: PatchBoardRequestDto, boardNumber:number, email:string): Promise<PatchBoardResponseDto> {
    
    const isExistUser = await this.userRepository.existsByEmail(email);
    if(!isExistUser) PatchBoardResponseDto.noExistUser();
    
    const boardEntity = await this.boardRepository.findByBoardNumber(boardNumber);
    if(!boardEntity) PatchBoardResponseDto.noExistBoard();

    const { writerEmail } = boardEntity;
    const isWriter = writerEmail === email;
    if(!isWriter) PatchBoardResponseDto.noPermission();

    boardEntity.title = dto.title;
    boardEntity.content = dto.content;
    await this.boardRepository.save(boardEntity);

    await this.imageRepository.deleteByBoardNumber(boardNumber);

    const { boardImageList } = dto;
    const imageEntities = this.imageRepository.createAll(boardImageList, boardNumber);
    await this.imageRepository.saveAll(imageEntities);

    return PatchBoardResponseDto.success();

  }

}
